x-pack/plugin/ml/log-structure-finder/src/main/java/org/elasticsearch/xpack/ml/logstructure/LogFileStructure.java

```diff
@@ -0,0 +1,614 @@
+    static final ParseField HAS_HEADER_ROW = new ParseField("has_header_row");
+    static final ParseField SEPARATOR = new ParseField("separator");
+    static final ParseField SHOULD_TRIM_FIELDS = new ParseField("should_trim_fields");
+    static final ParseField GROK_PATTERN = new ParseField("grok_pattern");
+    static final ParseField TIMESTAMP_FIELD = new ParseField("timestamp_field");
+    static final ParseField TIMESTAMP_FORMATS = new ParseField("timestamp_formats");
+    static final ParseField NEED_CLIENT_TIMEZONE = new ParseField("need_client_timezone");
+    static final ParseField MAPPINGS = new ParseField("mappings");
+    static final ParseField EXPLANATION = new ParseField("explanation");
+
+    public static final ObjectParser<Builder, Void> PARSER = new ObjectParser<>("log_file_structure", false, Builder::new);
+
+    static {
+        PARSER.declareInt(Builder::setNumLinesAnalyzed, NUM_LINES_ANALYZED);
+        PARSER.declareInt(Builder::setNumMessagesAnalyzed, NUM_MESSAGES_ANALYZED);
+        PARSER.declareString(Builder::setSampleStart, SAMPLE_START);
+        PARSER.declareString(Builder::setCharset, CHARSET);
+        PARSER.declareBoolean(Builder::setHasByteOrderMarker, HAS_BYTE_ORDER_MARKER);
+        PARSER.declareString((p, c) -> p.setFormat(Format.fromString(c)), STRUCTURE);
+        PARSER.declareString(Builder::setMultilineStartPattern, MULTILINE_START_PATTERN);
+        PARSER.declareString(Builder::setExcludeLinesPattern, EXCLUDE_LINES_PATTERN);
+        PARSER.declareStringArray(Builder::setInputFields, INPUT_FIELDS);
+        PARSER.declareBoolean(Builder::setHasHeaderRow, HAS_HEADER_ROW);
+        PARSER.declareString((p, c) -> p.setSeparator(c.charAt(0)), SEPARATOR);
+        PARSER.declareBoolean(Builder::setShouldTrimFields, SHOULD_TRIM_FIELDS);
+        PARSER.declareString(Builder::setGrokPattern, GROK_PATTERN);
+        PARSER.declareString(Builder::setTimestampField, TIMESTAMP_FIELD);
+        PARSER.declareStringArray(Builder::setTimestampFormats, TIMESTAMP_FORMATS);
+        PARSER.declareBoolean(Builder::setNeedClientTimezone, NEED_CLIENT_TIMEZONE);
+        PARSER.declareObject(Builder::setMappings, (p, c) -> new TreeMap<>(p.map()), MAPPINGS);
+        PARSER.declareStringArray(Builder::setExplanation, EXPLANATION);
+    }
+
+    private final int numLinesAnalyzed;
+    private final int numMessagesAnalyzed;
+    private final String sampleStart;
+    private final String charset;
+    private final Boolean hasByteOrderMarker;
+    private final Format format;
+    private final String multilineStartPattern;
+    private final String excludeLinesPattern;
+    private final List<String> inputFields;
+    private final Boolean hasHeaderRow;
+    private final Character separator;
+    private final Boolean shouldTrimFields;
+    private final String grokPattern;
+    private final List<String> timestampFormats;
+    private final String timestampField;
+    private final boolean needClientTimezone;
```

Comment:
Because it makes sense to include this in the output in all cases, whereas for some of the other boolean flags they only make sense when a particular high level structure has been found.  For example, `has_header_row` does not make sense at all for JSON.